docs/usor/intro-dp.md

```diff
@@ -81,24 +54,195 @@ probleme a cărora aparțin ele.
+        if (dp[poz] != -1) {
+            return dp[poz];
+        }
+        dp[poz] = min(solve(poz-1, v, dp) + abs(v[poz-1] - v[poz]), solve(poz-2, v, dp) + abs(v[poz-2] - v[poz]));
+        return dp[poz];
+    }
+
+    int main() {
+        int n;
+        cin >> n;
+        
+        vector<int> v(n+1);
+        vector<int> dp(n+1, -1);
+        
+        for (int i = 1; i <= n; i++) {
+            cin >> v[i];
+        }
+        
+        cout << solve(n, v, dp) << '\n';
+        return 0;
+    }
+    ```
+
+=== "Iterativ"
+
+    ```cpp
+    #include <iostream>
+    #include <vector>
+
+    using namespace std;
+
+    int main() {
+        int n;
+        cin >> n;
+        
+        vector<int> v(n+1);
+        vector<int> dp(n+1);
+        
+        for (int i = 1; i <= n; i++) {
+            cin >> v[i];
+        }
+        for (int i = 2; i <= n; i++) {
+            if (i == 2) {
+                dp[i] = abs(v[i] - v[i-1]);
+            }
+            else {
+                dp[i] = min(dp[i-1] + abs(v[i] - v[i-1]), dp[i-2] + abs(v[i] - v[i-2]));
+            }
+        }
+        
+        cout << dp[n] << '\n';
+        return 0;
+    }
+    ```
+
+<!-- Second Problem -->
+
+### Problema [Frog 2](https://atcoder.jp/contests/dp/tasks/dp_b)
+La fel ca la problema precedentă, trebuie să aflăm costul minim de care are nevoie o broască pentru a ajunge de pe prima piatră pe ultima, dacă se poate deplasa cu cel mult $k$ pași la un moment dat.
+
+Spre deosebire de problema precedentă, pentru a calcula $dp[i]$, ne vom folosi de informațiile din enunț, recurența de la Frog 1 va fi extinsă pentru a acoperi $k$ poziții. 
+
+$$
+dp[i] = min(dp[i-x] + abs(v[i] - v[i-x])), \forall x \leq k
+$$
+
+Cazul de bază constă în faptul că $dp[1] = 0$ și $dp[2] = abs(v[1] - v[2])$, iar complexitatea acestei soluții este $O(n \cdot k)$. 
+
+
+
+=== "Recursiv"
+
+    ```cpp
+    #include <iostream>
+    #include <vector>
+
+    using namespace std;
+
+    int solve(int poz, int diff, const vector<int> &v, vector<int> &dp) {
+        if (poz == 1) {
+            return 0;
+        }
+        if (dp[poz] != -1) {
+            return dp[poz];
+        }
+        dp[poz] = (1<<30); // 2^30
+        for (int i = 1; i <= diff; i++) {
+            if (poz - i >= 1) {
+                dp[poz] = min(dp[poz], solve(poz-i, diff, v, dp) + abs(v[poz] - v[poz-i]));
+            }
+        }
+        return dp[poz];
+    }
+    int main() {
+        int n, k;
+        cin >> n >> k;
+        
+        vector<int> v(n+1);
+        vector<int> dp(n+1, -1);
+        
+        for (int i = 1; i <= n; i++) {
+            cin >> v[i];
+        }
+        
+        cout << solve(n, k, v, dp) << '\n';
+        return 0;
+    }
+    ```
+=== "Iterativ"
+
+    ```cpp
+    #include <iostream>
+    #include <vector>
+
+    using namespace std;
+
+    int main() {
+        int n, k;
+        cin >> n >> k;
+        
+        vector<int> v(n+1);
+        vector<int> dp(n+1);
```

Comment:
Alternativ:
```cpp
vector<int> v(n+1), dp(n+1, 1 << 30);
```
și scoți atribuirea din for-ul cu j

Reply:
Fac schimbarea dar cum am zis și înainte, tot cred că afectă complexitatea timpului adăugând o constantă extra. Da este mai ok pentru citibilitate and stuff și din acest motiv fac schimbarea.

Reply:
> Fac schimbarea dar cum am zis și înainte, tot cred că afectă complexitatea timpului adăugând o constantă extra. Da este mai ok pentru citibilitate and stuff și din acest motiv fac schimbarea.

Vezi ce am vorbit la inițializare și nu mai da 🧢, vino la mine cu un benchmark sau ceva de pe Compiler Explorer cu -O2/O3/Ofast și zi-mi cum arată output-ul față de O0.